### PR TITLE
Generate txhashset archives on 720 block intervals.

### PR DESCRIPTION
**Pull request by cadmuspeverell**
This is the first of several PRs to resolve https://github.com/mimblewimble/grin/issues/2740. With this PR, txhashset archives will only be generated for blocks 250, 500, 750, etc. This inadvertently solves https://github.com/mimblewimble/grin/issues/2806 as well.

Marked as WIP to encourage review and discussion while I write tests.

### CODE REVIEW

**Comment by garyyu** [comment]
Finally, all 3 brothers are here 😃 👍

**Comment by antiochp** [comment]
Hello brother.

**Comment by cadmuspeverell** [comment]
This is ready for review whenever you get a chance @antiochp

File: p2p/src/protocol.rs [hunk 1]
```diff
@@ -251,15 +251,17 @@ impl MessageHandler for Protocol {
 					sm_req.hash, sm_req.height
 				);
 
-				let txhashset = self.adapter.txhashset_read(sm_req.hash);
```
Comment (antiochp) [inline p2p/src/protocol.rs]: We need to think this through a bit. Our peer sends us a height and hash (the `sm_req`) but we are now effectively ignoring this and just sending the most recent snapshot.

I wonder if we need to take the height send by our peer and quantize this to the nearest previous snapshot (i.e. the closest `720` snapshot to that height).

Alternatively maybe we should be ignoring the values passed in by our peer - all we care about is serving the most recent snapshot?

Comment (cadmuspeverell) [reply p2p/src/protocol.rs]: I thought about it for a while, but I don't think we would want to provide them any other txhashset. I considered verifying that the requested header is within an expected range, but it seemed unlikely enough to not justify any additional complexity.

**Comment by antiochp** [comment]
I need to think through the horizon behavior here a bit more but looks good so far.

**Comment by cadmuspeverell** [comment]
> I need to think through the horizon behavior here a bit more but looks good so far.

Sure, take your time.  This does effectively change the horizon temporarily for newly synced nodes.  Once most nodes have this change, it might make sense to just make txhashset_archive_header the new cut_through_horizon.  Perhaps as part of the first hard fork?

**Comment by antiochp** [comment]
Currently we make some guarantees/assumptions about what blocks are available from peers.

We maintain 7 days of full block history based on `global::cut_through_horizon()`.
But on initial sync we ask for a txhashset based on 48 hours with `global::state_sync_threshold()`.

So every running node has at least 48 hours of full blocks (after successful sync).
And every running node has up to 7 days of full blocks based on compaction/pruning.

One of the edge cases we need to account for here is when we ask for full blocks during fast sync from another node that has recently sync'd. i.e. Other nodes (possible a majority of nodes on the network) are only guaranteed to have blocks from 48 hours ago. Currently this is _exactly_ 48 hours ago.

So there are a couple of conflicting requirements here pulling in competing directions -

* If we ask for a txhashset older than 48 hours then we risk not being able to reliably sync full blocks from that point onwards.
* If we ask for a txhashset younger than 48 hours then _we_ can (selfishly) sync but we are unable to provide 48 hours worth of blocks to other peers who may also by sync'ing.

So currently we are kind of stuck doing the same thing as all other peers and setting our sync_threshold to 48 hours.

If we were to deploy a node with different behaviour by for example using more of a "step function" in terms of 740 block periods then we risk being a bad citizen with respect to these conflicting requirements. We either have too few blocks locally (temporarily) or we are asking for too many blocks and cannot reliably retrieve them (if our peers are also syncing).

**Comment by cadmuspeverell** [comment]
> We  either have too few blocks locally (temporarily) or we are asking for  too many blocks and cannot reliably retrieve them (if our peers are also syncing).

The way it's coded in this PR, it would be the latter.  This was coded that way intentionally, for the reasons you listed above.  Although initial sync might be delayed very slightly if it requests blocks from a peer that just synced, it shouldn't be a large problem because the significant majority of peers would have more than 48 hours worth of blocks.

**Comment by garyyu** [comment]
@cadmuspeverell  thanks for your work and how are you doing?

I would like to take a look and test this PR and could you please solve the conflict?  thanks.

**Comment by antiochp** [comment]
Q) If we only produce a new txhashset every 12 hours - do we now reuse one already generated if we are asked for it more than once in a 12 hour period?

**Comment by cadmuspeverell** [comment]
> Q) If we only produce a new txhashset every 12 hours - do we now reuse one already generated if we are asked for it more than once in a 12 hour period?
> 
> I don't remember if we implemented de-duping behavior before (we had a smaller window before).

Yep!  That logic was already there - and seemed to work well when I was doing my testing.  I'll resolve that darned conflict now and then maybe @garyyu can try it out to make sure I didn't miss anything?  :smile_cat:

**Comment by garyyu** [comment]
After my 2nd reading everything is good on my view, and thanks @cadmuspeverell for this 1st PR 👍 and a long review :-)

**Comment by antiochp** [comment]
Looks like this PR was merged into `master` after the release of `1.1.0`.
Can we revert this on `master` and retarget it for the `2.0.0` branch (its tagged as `2.x.x` milestone). 

If we do need to push `1.1.x` for any reason we will do this from `master` and we only want to include critical fixes.

I'm not 100% sure if this is the approach we will always take but we have some tight time constraints between `1.1.0` and `2.0.0` and we decided to keep `master` clean for a `1.1.1` release if necessary.

**Comment by garyyu** [comment]
> its tagged as 2.x.x milestone

but it was tagged as 1.1.1 milestone before, you can see above history.
And I don't see any risk for this PR in 1.1.1, and we didn't yet plan a release date about 1.1.1, did we?

**Comment by yeastplume** [comment]
@garyyu There is no 1.1.1 milestone anymore, in recent meetings we've collectively decided that the next release will be 2.0.0, and that there will only be a 1.1.1 if there is a critical issue uncovered that warrants it. Also, given the time constraints we should only be pushing PRs that are hard-fork related to the branch, and labeling any other fixes or changes as post 2.0.0 (i've created a 2.x.x milestone for that)

Sorry this wasn't communicated properly as there have been one or two extra meeting schedules over the past days, but we really feel it's important that everyone stick to this plan for the next few weeks.

**Pull request merged**